lib/ecto_adapters_dynamodb/query.ex

```diff
@@ -90,58 +100,89 @@ defmodule Ecto.Adapters.DynamoDB.Query do
           end)
         else
           do_fetch_recursive.(query)
         end
       :scan ->
         maybe_scan(table, search, opts)
     end
+  end
 
-    results
+  @spec filter(dynamo_response,  non_indexed_filters, String.t) :: dynamo_response | %{}
+  defp filter(%{"Item" => item} = result, non_indexed_filters, _) do
+    if passes_filter?(item, non_indexed_filters),
+      do: result,
+      else: %{}
   end
+  defp filter(%{@responses_element_key => items} = result, non_indexed_filters, table) do
+    filtered_results = Enum.filter(items[table], &passes_filter?(&1, non_indexed_filters))
+    put_in(result, [@responses_element_key, table], filtered_results)
+  end
+
 
-  # If a primary key query has additional search clauses that are not reflected by the indexes,
-  # we may need to use a Dynamo query instead of get_item in order to apply filters.
-  defp should_query?(_indexes, []), do: false
-  defp should_query?(indexes, [{logical_op, search_clauses} | additional_search_clauses]) when logical_op in @logical_ops,
-    do: should_query?(indexes, search_clauses) or should_query?(indexes, additional_search_clauses)
-  defp should_query?(indexes, [{field, _} | search_clauses]) do
-    if field not in indexes,
-      do: true,
-      else: should_query?(indexes, search_clauses)
+  # @spec passes_filter?(dynamo_response, non_indexed_filters) :: boolean()
+  defp passes_filter?(item, [{ logical_op, [ deeper_clauses, filter ] }]) do
+    case logical_op do
+      :and -> evaluate_filter_expression(item, filter) and passes_filter?(item, deeper_clauses)
+      :or -> evaluate_filter_expression(item, filter) or passes_filter?(item, deeper_clauses)
+    end
   end
+  defp passes_filter?(item, [{ _, [ filter ] }]) do
+    evaluate_filter_expression(item, filter)
+  end
+  defp passes_filter?(item, deeper_clauses) do
+    passes_filter?(item , [ deeper_clauses ])
+  end
+
+
+  # @spec evaluate_filter_expression(decoded_terms, filter_clause) :: boolean()
+  defp evaluate_filter_expression(item, { field, filter_clause }) when is_map(item) do
+    case Map.get(item, field) do
+      nil -> evaluate_filter_expression(nil, filter_clause)
+      encoded_value ->
+        encoded_value
+        |> Decoder.decode()
+        |> evaluate_filter_expression(filter_clause)
+    end
+  end
+  defp evaluate_filter_expression(value, { filter_val, op }) when op in [ :==, :<, :<=, :>, :>= ],
+    do: apply(Kernel, op, [value, filter_val])
+  defp evaluate_filter_expression(value, { _, :is_nil }),
+    do: is_nil(value)
+  defp evaluate_filter_expression(value, { filter_val, :in }),
+    do: value in filter_val
+  defp evaluate_filter_expression(value, { [ range_start, range_end ], :between }),
+    do: value >= range_start and value <= range_end
+  defp evaluate_filter_expression(value, { filter_val, :begins_with }) do
+    {:ok, reg} = Regex.compile("^#{filter_val}.*")
```

Comment:
`String.starts_with?` ?

Reply:
Somehow, I missed that in the docs - that'll be fixed in a subsequent push.